Solving `move`
-------------

Turns out there is a solution to the `move` function ; it's called a Result.
See https://www.reddit.com/r/elm/comments/4769nd/modeling_handling_of_invalid_arguments/d0arxd3

```elm
type Result =
      Ok
    | Err
```

So changing the signature to :

```elm
move : ( Int, Int ) -> Cell -> Board -> Result BadMove Board
```

Maybes it possible to use the `Result.andThen` function, to chain calls :

```elm
alreadyTaken : Result Ttt.BadMove Ttt.Board
alreadyTaken =
  Ttt.move ( 2, 2 ) Ttt.X Ttt.emptyBoard
    `Result.andThen` Ttt.move ( 2, 2 ) Ttt.O
```

It forwards calls, and propagates errors if any is found.

I don't know what happens if you try to chain calls that return different types of errors...


```elm
type FooError
  = TooFoo
  | NotFoo


type BarError
  = TooBar
  | NotBar


foo : Int -> Result FooError Int
foo i =
  if (i < 0) then
    Err NotFoo
  else if (i > 10) then
    Err TooFoo
  else
    Ok i


bar : Int -> Result BarError Int
bar i =
  if (i < 3) then
    Err NotBar
  else if (i > 5) then
    Err TooBar
  else
  Ok i
```

You can chain calls of the same function (or anything with the right types)

```elm
ok : Result FooError Int
ok =
  (foo 1) `Result.andThen` foo
```

But try mixing types, like this : 

```elm
-- Will fail !           
what = (foo 1) `Result.andThen` bar
```

And you get one of the *brilliants* error messages from Elm compiler.

```
The right argument of `andThen` is causing a type mismatch.
           
107â”‚         foo 1) `Result.andThen` bar
                                     ^^^

`andThen` is expecting the right argument to be a:
                                
    Int -> Result FooError a

But the right argument is:

Int -> Result BarError Int
```

So, using results it is.


